aptos-move/aptos-replay-benchmark/src/diff.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use aptos_types::{
+    contract_event::ContractEvent,
+    state_store::state_key::StateKey,
+    transaction::{ExecutionStatus, TransactionOutput},
+    write_set::{WriteOp, WriteSet},
+};
+use claims::assert_ok;
+use std::collections::BTreeMap;
+
+/// Different parts of [TransactionOutput] that can be different:
+///   1. gas used,
+///   2. status (must be kept since transactions are replayed),
+///   3. events,
+///   4. writes.
+/// Note that fine-grained comparison allows for some differences to be okay, e.g., using more gas
+/// implies that the fee statement event, the account balance of the fee payer, and the total token
+/// supply are different.
+enum Diff {
+    GasUsed {
+        left: u64,
+        right: u64,
+    },
+    ExecutionStatus {
+        left: ExecutionStatus,
+        right: ExecutionStatus,
+    },
+    Event {
+        left: Option<ContractEvent>,
+        right: Option<ContractEvent>,
+    },
+    WriteSet {
+        left: Option<(StateKey, WriteOp)>,
+        right: Option<(StateKey, WriteOp)>,
+    },
+}
+
+/// Holds all differences for a pair of transaction outputs.
+pub(crate) struct TransactionDiff {
+    diffs: Vec<Diff>,
```

Comment:
can have read diffs here separately, if we do the way suggested earlier - could be just statekeys.

Reply:
I think you still want to have write ops to be able to not count as diffs certain things: e.g. creation and modification do not matter as long as bytes are the same, etc.

Reply:
or do you mean have gas_diff, write_set_diff, etc?

Reply:
I thought about that but as you do now seems more useful, I was saying we can have this be output_diff and we can have read / input diffs as well. I mean a diff in terms of what was accessed.

Reply:
I see! Then we need to get the read set for a txn, we only have for pre-block? Ideally we can use it from Block-STM, but that info is private there right now.

Reply:
oh the read set capturing stateview does it for everything together, fair point. We can indeed get that differently, can have a todo (and figure out the best way when addressing that)

Reply:
Added a todo for read-sets